yggdrasilffi/src/lib.rs

```diff
@@ -56,257 +63,241 @@ impl From<serde_json::Error> for FFIError {
+    CString::new(json_string).unwrap().into_raw()
+}
+
+/// Instantiates a new engine. Returns a pointer to the engine.
+///
 /// # Safety
 ///
-/// The caller is responsible for ensuring the pointer to the engine is a valid pointer,
-/// a passed null pointer is a no-op but an invalid pointer is undefined behavior.
-/// This function must be called correctly to deallocate the memory allocated for the engine in
-/// the engine_new function, failure to do so will result in a leak.
+/// The caller is responsible for freeing the allocated memory. This can be done by calling
+/// `free_engine` and passing in the pointer returned by this method. Failure to do so will result in a leak.
 #[no_mangle]
-pub unsafe extern "C" fn engine_free(ptr: *mut c_void) {
-    if ptr.is_null() {
+pub extern "C" fn new_engine() -> *mut c_void {
+    let engine = EngineState::default();
+    Box::into_raw(Box::new(engine)) as *mut c_void
+}
+
+/// Frees the memory allocated for the engine.
+///
+/// # Safety
+///
+/// The caller is responsible for ensuring the argument is a valid pointer.
+/// Null pointers will result in a no-op, but any invalid pointers will result in undefined behavior.
+/// These pointers should not be dropped for the lifetime of this function call.
+///
+/// This function must be called correctly in order to deallocate the memory allocated for the engine in
+/// the `new_engine` function. Failure to do so will result in a leak.
+#[no_mangle]
+pub unsafe extern "C" fn free_engine(engine_ptr: *mut c_void) {
+    if engine_ptr.is_null() {
         return;
     }
-    unsafe {
-        drop(Box::from_raw(ptr as *mut EngineState));
-    }
+    drop(Box::from_raw(engine_ptr as *mut EngineState));
 }
 
-/// Takes a JSON string representing a set of toggles and returns a JSON string representing
-/// the metrics that should be sent to the server
+/// Takes a JSON string representing a set of toggles. Returns a JSON string representing
+/// the metrics that should be sent to the server.
+///
 /// # Safety
 ///
-/// The caller is responsible for ensuring the pointer to the engine is a valid pointer
-/// and that the engine is not dropped while the pointer is still in use.
+/// The caller is responsible for ensuring all arguments are valid pointers.
+/// Null pointers will result in an error message being returned to the caller,
+/// but any invalid pointers will result in undefined behavior.
+/// These pointers should not be dropped for the lifetime of this function call.
 #[no_mangle]
-pub unsafe extern "C" fn engine_take_state(ptr: *mut c_void, json: *const c_char) -> *const c_char {
-    let response: Response<()> = take_state(ptr, json).into();
-    let json = serde_json::to_string(&response).unwrap();
-    CString::new(json).unwrap().into_raw()
-}
+pub extern "C" fn take_state(engine_ptr: *mut c_void, json_ptr: *const c_char) -> *const c_char {
+    let result: Result<Option<()>, FFIError> = (|| {
```

Comment:
Why the immediately executed closure? Is this so you can get the error handling with the single function? Not against it, just new to me as an idea

Reply:
Well, I wanted to standardize our methods, since they mostly consist of grabbing the values from the pointers we pass in and then returning a response based on the internal method we execute - Basically `Result<Option<T>, FFIError>`. With this in mind, we could have an auxiliary method like `result_to_json_ptr` that would convert our response to a pointer we could return. I also wanted to avoid having a separate auxiliary method for each public method to improve consistency, readability and maintainability.

So this is what I came up with. It feels clean to me but maybe it's actually more readable and maintainable to have separate methods? 🤔 I'm open to suggestions here!

Reply:
I can get behind this. There's a small cost to wrapping stuff in closures like this but given what the rest of the file is doing, it shouldn't be noticeable. It makes things nicer to read and cleaner

I'm okay with responses for everything except the two metrics methods. A caller can't respond to a failure there in any meaningful way so it makes sense for them to be returnless. Everything else I think now has a standard response format

Reply:
You're talking about:
 - `count_toggle`
 - `count_variant`
 
Correct? What should happen in case we have an error there (e.g. one of the pointers is null)? From an interface perspective, I think it makes sense that we still let the caller know, even if it decides to just log it.

I'll merge this for now but I'm interested to know what you have in mind.